File: analytics/src/main/java/com/bluetriangle/analytics/BlueTriangleConfiguration.java
package com.bluetriangle.analytics;

import android.util.Log;

import java.util.concurrent.TimeUnit;
import java.util.logging.Level;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public class BlueTriangleConfiguration {

    public static final String DEFAULT_TRACKER_URL = "https://d.btttag.com/analytics.rcv";
    public static final String DEFAULT_ERROR_REPORTING_URL = "https://d.btttag.com/err.rcv";

    /**
     * The Blue Triangle site ID
     */
    @Nullable private String siteId;

    /**
     * Global User ID
     */
    @Nullable private String globalUserId;

    /**
     * Session ID
     */
    @Nullable private String sessionId;

    /**
     * the application's name
     */
    @Nullable private String applicationName;

    @Nullable private String userAgent;

    private boolean debug = false;
    private int debugLevel = Log.DEBUG;
    @Nullable private Logger logger = NoOpLogger.getInstance();

    @NonNull private String trackerUrl = DEFAULT_TRACKER_URL;
    @NonNull private String errorReportingUrl = DEFAULT_ERROR_REPORTING_URL;

    /**
     * Cache directory path
     */
    private String cacheDirectory = null;

    /**
     * Max items in the cache
     */
    private int maxCacheItems = 100;

    /**
     * Max attmepts to resend a payload
     */
    private int maxAttempts = 3;

    /**
     * the instance to cache payloads
     */
    private PayloadCache payloadCache = null;

    private boolean trackCrashesEnabled = false;
    private boolean performanceMonitorEnabled = true;
    private long performanceMonitorIntervalMs = TimeUnit.SECONDS.toMillis(1);

    @Nullable
    public String getSiteId() {
        return siteId;
    }

    public void setSiteId(@Nullable String siteId) {
        this.siteId = siteId;
    }

    @Nullable
    public String getSessionId() {
        return sessionId;
    }

    public void setSessionId(@Nullable String sessionId) {
        this.sessionId = sessionId;
    }

    @Nullable
    public String getGlobalUserId() {
        return globalUserId;
    }

    public void setGlobalUserId(@Nullable String globalUserId) {
        this.globalUserId = globalUserId;
    }

    @Nullable
    public String getUserAgent() {
        return userAgent;
    }

    public void setUserAgent(@Nullable String userAgent) {
        this.userAgent = userAgent;
    }

    @Nullable
    public String getApplicationName() {
        return applicationName;
    }

    public void setApplicationName(@Nullable String applicationName) {
        this.applicationName = applicationName;
    }

    /**
     * get the cache directory path
     *
     * @return cache directory path
     */
    public String getCacheDirectory() {
        return cacheDirectory;
    }

    /**
     * Set the cache directory path
     *
     * @param cacheDirectory path to cache director
     */
    public void setCacheDirectory(String cacheDirectory) {
        this.cacheDirectory = cacheDirectory;
    }

    public int getMaxCacheItems() {
        return maxCacheItems;
    }

    public void setMaxCacheItems(int maxCacheItems) {
        this.maxCacheItems = maxCacheItems;
    }

    public int getMaxAttempts() {
        return maxAttempts;
    }

    public void setMaxAttempts(int maxAttempts) {
        this.maxAttempts = maxAttempts;
    }

    public PayloadCache getPayloadCache() {
        if (payloadCache == null) {
            payloadCache = new PayloadCache(this);
        }
        return payloadCache;
    }

    public void setPayloadCache(PayloadCache payloadCache) {
        this.payloadCache = payloadCache;
    }

    public boolean isDebug() {
        return debug;
    }

    public void setDebug(boolean debug) {
        this.debug = debug;
    }

    public int getDebugLevel() {
        return debugLevel;
    }

    public void setDebugLevel(int debugLevel) {
        this.debugLevel = debugLevel;
    }

    @NonNull
    public Logger getLogger() {
        if (logger == null) {
            if (debug) {
                logger = new AndroidLogger(debugLevel);
            } else {
                logger = NoOpLogger.getInstance();
            }
        }
        return logger;
    }

    public void setLogger(@NonNull Logger logger) {
        this.logger = logger;
    }

    @NonNull
    public String getTrackerUrl() {
        return trackerUrl;
    }

    public void setTrackerUrl(@NonNull String trackerUrl) {
        this.trackerUrl = trackerUrl;
    }

    @NonNull
    public String getErrorReportingUrl() {
        return errorReportingUrl;
    }

    public void setErrorReportingUrl(@NonNull String errorReportingUrl) {
        this.errorReportingUrl = errorReportingUrl;
    }

    public boolean isPerformanceMonitorEnabled() {
        return performanceMonitorEnabled;
    }

    public void setPerformanceMonitorEnabled(boolean performanceMonitorEnabled) {
        this.performanceMonitorEnabled = performanceMonitorEnabled;
    }

    public long getPerformanceMonitorIntervalMs() {
        return performanceMonitorIntervalMs;
    }

    public void setPerformanceMonitorIntervalMs(long performanceMonitorIntervalMs) {
        this.performanceMonitorIntervalMs = performanceMonitorIntervalMs;
    }

    public boolean isTrackCrashesEnabled() {
        return trackCrashesEnabled;
    }

    public void setTrackCrashesEnabled(boolean trackCrashesEnabled) {
        this.trackCrashesEnabled = trackCrashesEnabled;
    }
}

File: analytics/src/main/java/com/bluetriangle/analytics/BlueTriangleConfiguration.kt
package com.bluetriangle.analytics

import android.Manifest
import android.util.Log
import androidx.annotation.RequiresPermission
import java.util.concurrent.TimeUnit

class BlueTriangleConfiguration {
    /**
     * The Blue Triangle site ID
     */
    var siteId: String? = null

    /**
     * Global User ID
     */
    var globalUserId: String? = null

    /**
     * Session ID
     */
    var sessionId: String? = null

    /**
     * the application's name
     */
    var applicationName: String? = null

    /**
     * SDK's user agent
     */
    var userAgent: String? = null

    /**
     * Percentage of sessions for which network calls will be captured. A value of `0.05` means that 5% of session's network requests will be tracked.
     * A value of `0.0` means that no network requests will be captured for the session, a value of `1.0` will track all network requests for a session.
     */
    var networkSampleRate = 0.05
        set(networkSampleRate) {
            field = Math.min(Math.max(networkSampleRate, 0.0), 1.0)
            shouldSampleNetwork = Utils.shouldSample(this.networkSampleRate)
        }

    /**
     * if network requests should be captured
     */
    var shouldSampleNetwork = false

    var isDebug = false
    var debugLevel = Log.DEBUG
    var logger: Logger? = null
        get() {
            if (field == null) {
                field = AndroidLogger(debugLevel)
            }
            return field
        }

    var trackerUrl = DEFAULT_TRACKER_URL

    var errorReportingUrl = DEFAULT_ERROR_REPORTING_URL

    var networkCaptureUrl = DEFAULT_NETWORK_CAPTURE_URL

    /**
     * Cache directory path
     */
    var cacheDirectory: String? = null

    /**
     * Max items in the cache
     */
    var maxCacheItems = 100

    /**
     * Max attempts to resend a payload
     */
    var maxAttempts = 3

    /**
     * the instance to cache payloads
     */
    var payloadCache: PayloadCache? = null
        get() {
            if (field == null) {
                field = PayloadCache(this)
            }
            return field
        }

    /**
     * Enable or Disable automatic crash detection and reporting
     */
    var isTrackCrashesEnabled = false

    /**
     * Enable or disable monitoring and reporting performance metrics
     */
    var isPerformanceMonitorEnabled = false

    /**
     * Enable or disable memory warning detection and reporting
     */
    var isMemoryWarningEnabled = false

    /**
     * Set the sampling interval for performance monitoring in milliseconds
     */
    var performanceMonitorIntervalMs = TimeUnit.SECONDS.toMillis(1)
        set(value) {
            field = value.coerceAtLeast(500L)
        }

    /**
     * Enable or disable ANR detection and sending reports to the server.
     */
    var isTrackAnrEnabled: Boolean = false

    /**
     * time interval for ANR warning based on track ANR is enabled or disabled, default to 5 seconds, minimum is 3 second, if set less then minimum allowed set value is ignored
     */
    var trackAnrIntervalSec = Constants.ANR_DEFAULT_INTERVAL

    /**
     * Enable or disable automatic tracking and reporting of page views for Activity and Fragments.
     * For Jetpack compose use BttTimerEffect
     */
    var isScreenTrackingEnabled: Boolean = false

    /**
     * Enable or disable tracking and reporting launch time.
     * Supported on API Level 29 and above
     */
    var isLaunchTimeEnabled: Boolean = false

    /**
     * Track the network state during Timer, Network request and errors. States include wifi, cellular, ethernet and offline.
     * Default value is false.
     * Requires host app to have ACCESS_NETWORK_STATE permission.
     */
    @RequiresPermission(Manifest.permission.ACCESS_NETWORK_STATE)
    var isTrackNetworkStateEnabled: Boolean = false

    companion object {
        const val DEFAULT_TRACKER_URL = "https://d.btttag.com/analytics.rcv"
        const val DEFAULT_ERROR_REPORTING_URL = "https://d.btttag.com/err.rcv"
        const val DEFAULT_NETWORK_CAPTURE_URL = "https://d.btttag.com/wcdv02.rcv"
    }
}